packages/amplify-graphql-transformer-core/src/transformation/utils.ts

```diff
@@ -0,0 +1,180 @@
+import {
+  DirectiveNode,
+  TypeDefinitionNode,
+  FieldDefinitionNode,
+  InputValueDefinitionNode,
+  EnumValueDefinitionNode,
+  DirectiveDefinitionNode,
+  TypeSystemDefinitionNode,
+  Kind,
+} from 'graphql';
+import { TransformerPluginProvider, TransformerPluginType } from '@aws-amplify/graphql-transformer-interfaces';
+
+export function makeSeenTransformationKey(
+  directive: DirectiveNode,
+  type: TypeDefinitionNode,
+  field?: FieldDefinitionNode | InputValueDefinitionNode | EnumValueDefinitionNode,
+  arg?: InputValueDefinitionNode,
+  index?: number,
+): string {
+  let key = '';
+  if (directive && type && field && arg) {
+    key = `${type.name.value}.${field.name.value}.${arg.name.value}@${directive.name.value}`;
+  }
+  if (directive && type && field) {
+    key = `${type.name.value}.${field.name.value}@${directive.name.value}`;
+  } else {
+    key = `${type.name.value}@${directive.name.value}`;
+  }
+  if (index !== undefined) {
+    key += `[${index}]`;
+  }
+  return key;
+}
+
+/**
+ * If this instance of the directive validates against its definition return true.
+ * If the definition does not apply to the instance return false.
+ * @param directive The directive definition to validate against.
+ * @param nodeKind The kind of the current node where the directive was found.
+ */
+export function matchDirective(definition: DirectiveDefinitionNode, directive: DirectiveNode, node: TypeSystemDefinitionNode) {
+  if (!directive) {
+    return false;
+  }
+  if (definition.name.value !== directive.name.value) {
+    // The definition is for the wrong directive. Do not match.
+    return false;
+  }
+  let isValidLocation = false;
+  for (const location of definition.locations) {
```

Comment:
does only one location have to be valid or all of them? This logic currently only checks that one is valid

(same in the validation logic for the subsequent methods as well)

Reply:
One of the valid location is good enough. This is making sure that the location where the directive is used is valid